Developer onboarding Guide - Closes #1082

frnco opened the pull request:
Spread through a few commits for easier Reviewing.

Basically there's
- Two docker-compose files (One for dev, one for test);
- A new .env.dev file for Development Environment Variables;
- Dockerfile installs sphinx because it is needed to run ts:configure;
- .dockerignore ignores the new docker-compose-test file;
- config/thinking_sphynx.yml has the needed values in the development section;
- run-tests.sh runs the same tests as Travis;
- CONTRIBUTING.md Contains a basic onboarding guide;
- README.md has a Link to CONTRIBUTING.md

juanj commented on CONTRIBUTING.md:
````diff
@@ -1,3 +1,116 @@
+```
+
+### Some things you want to keep in mind
+- The database as well as the Search index use volumes so your data is persisted across runs, uless you remove the containers, so the one-time setup really means one-time.
````
I think there's a typo here. Maybe 'unless' ?

frnco commented:
@juanj good catch, thanks for taking a look, I fixed it.

I also found another issue (I had rephrased some parts and forgot to remove an "it"), plus I added a part about signing the CLA.

wkonkel commented on run-tests.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+/app/bin/rake db:drop db:create spec:prepare
```
This file scares me because if somebody accidentally runs it, their local database gets dropped. Also, why are you running the specs separately and dropping databases rather than:
```/app/bin/bundle exec rspec --pattern spec/**/*_spec.rb``` or ```./bin/rake``` ?

frnco replied:
This one I won't argue, I completely agree and I am already working on it, it was the fastest and hackiest way to have instructions on Testing, but this will at the very least go into a Rake Task, but I'm still trying to make docker-compose run everything by itself. Les commands, less complexity, more happiness.

The reason for the many lines is, beside sending the PR sooner, `spec/vcr`, which fails when we run `rspec`, so I simply copied what Travis runs to get the discussion going.

Ideally we shouldn't use patterns, if a test is supposed to fail we're supposed to take it off or move it to its feature branch.

frnco replied:
Removed, finally. :D

frnco commented on CONTRIBUTING.md:
```diff
@@ -1,3 +1,119 @@
+
+First of all you need [Docker](https://www.docker.com/what-docker) and [Docker Compose](https://docs.docker.com/compose/install/) installed.
+
+## Environment
```
Not sure where it should go, I want to refactor it to a more appropriate place though. GitHub Wiki seems like a good idea.

frnco replied:
Moved to Wiki

frnco commented on CONTRIBUTING.md:
```diff
@@ -1,3 +1,119 @@
+
+Mainly, you need to create the Sphinx configuration file as well as Symlink it to its expected location before running the Search Daemon (Default command Sphinx image runs). You also have to generate the database and search indexes.
+
+0. Build container and start database.  
```
It's a trade-off. I like the Rails solution of having a separate repo and I considered nesting Docker Images which is cleaner that this, but both will take a while and we need to have some way for people to get up and running.

Considering that, the easiest way seems to be Docker, but obviously we can look at other options, the PR was so I could get feedback. With Docker though there's no need for worrying about Ruby Versions/RVM, no need to install PostGres or Sphinx, just install latest Docker/Docker-Compose and run 5 commands.

And I would actually like to take those 5 commands down to a single one, but decided to PR before that so we could discuss it.

Do you think I should change it to RVM/Sphinx/Database (PostGres or maybe SQLite since it's easier, but I have to check if it will work...).

frnco replied:
I'm taking a look at `foreman` and updated the PR with the improved workflow for Docker so we have a working alternative.

Foreman would also require manually managing versions and dependencies, which, after a few years of managing servers and using a Rolling Release Linux distro, I try to avoid at all costs. It may also lead to bugs due to differences in developer environments.

We could also provide both options, so each developer can decide. Do you think it's a good idea, or would you rather completely remove Docker?

frnco commented:
There's also the matter of seeding some data which I postponed, but it needs to be decided. Considering development may bring the DB to an Inconsistent or even broken state, it should be easier to drop and re-create and re-seed the database, but clear enough so people don't do it by mistake. I considered rake tasks, any opinions on that?

alphashuro commented:
@frnco Any updates on the seeding issue?
